File: packages/swift-server/Sources/Keychain/SecretMasking.swift
import CommonCrypto
import Foundation

// MARK: - Known Token Prefixes

private let knownPrefixes: [String] = [
    "github_pat_",
    "sk-ant-",
    "Bearer ",
    "ghp_",
    "gho_",
    "ghu_",
    "ghs_",
    "ghr_",
    "xoxb-",
    "xoxp-",
    "xoxa-",
    "xoxs-",
    "sk-",
    "pk-",
    "AKIA",
    "ABIA",
    "ACCA",
    "ASIA",
]

// Sorted longest-first for most-specific matching
private let sortedPrefixes = knownPrefixes.sorted { $0.count > $1.count }

private func detectPrefix(_ value: String) -> String {
    for p in sortedPrefixes {
        if value.hasPrefix(p) { return p }
    }
    return ""
}

// MARK: - HMAC-SHA256

private func hmacSHA256(key: String, message: String) -> [UInt8] {
    let keyData = Array(key.utf8)
    let messageData = Array(message.utf8)
    var result = [UInt8](repeating: 0, count: Int(CC_SHA256_DIGEST_LENGTH))
    CCHmac(
        CCHmacAlgorithm(kCCHmacAlgSHA256),
        keyData, keyData.count,
        messageData, messageData.count,
        &result
    )
    return result
}

private func toHex(_ bytes: [UInt8]) -> String {
    bytes.map { String(format: "%02x", $0) }.joined()
}

// MARK: - Public API

/// Produce a deterministic, format-preserving masked value.
///
/// `masked = prefix + hex(HMAC-SHA256(sessionId+secretName, realValue))`
/// truncated or repeated to match the original value length.
///
/// Length math operates on UTF-16 code units (`String.utf16.count`) rather
/// than grapheme clusters (`String.count`) to match the TypeScript reference
/// (`packages/shared-ts/src/secret-masking.ts`, which uses `.length`). For
/// ASCII inputs the two agree; for inputs containing emoji or surrogate
/// pairs they diverge and the cross-implementation mask vector contract
/// would break.
public func mask(sessionId: String, secretName: String, realValue: String) -> String {
    let prefix = detectPrefix(realValue)
    // UTF-16 code-unit length matches JS `String.length` semantics. Grapheme-
    // cluster `.count` would diverge from the TS reference for inputs
    // containing emoji or surrogate pairs and break the cross-implementation
    // mask vector contract. hex is pure ASCII so `.count` ≡ `.utf16.count`.
    let prefixUTF16Length = prefix.utf16.count
    let remainderUTF16Length = realValue.utf16.count - prefixUTF16Length

    let hmac = hmacSHA256(key: sessionId + secretName, message: realValue)
    var hex = toHex(hmac)

    // Repeat hex if remainder is longer than 64 hex chars
    while hex.count < remainderUTF16Length { hex += hex }
    let maskedRemainder = String(hex.prefix(remainderUTF16Length))

    return prefix + maskedRemainder
}

/// A pair of real ↔ masked secret values for scrubbing.
public struct SecretPair {
    public let realValue: String
    public let maskedValue: String

    public init(realValue: String, maskedValue: String) {
        self.realValue = realValue
        self.maskedValue = maskedValue
    }
}

/// Build a reusable scrubber that replaces every occurrence of any
/// `realValue` with its `maskedValue`.
///
/// Secrets are sorted longest-first to avoid partial-match clobbering.
public func buildScrubber(secrets: [SecretPair]) -> @Sendable (String) -> String {
    guard !secrets.isEmpty else { return { $0 } }

    let sorted = secrets.sorted { $0.realValue.count > $1.realValue.count }

    return { text in
        var result = text
        for pair in sorted {
            result = result.replacingOccurrences(of: pair.realValue, with: pair.maskedValue)
        }
        return result
    }
}

/// Domain glob matching.
///
/// - `*` matches any domain
/// - `api.github.com` matches exactly
/// - `*.github.com` matches `api.github.com`, `uploads.github.com`,
///   but NOT `github.com` itself
public func domainMatches(pattern: String, hostname: String) -> Bool {
    let p = pattern.lowercased()
    let h = hostname.lowercased()

    // Bare wildcard: allow any domain
    if p == "*" { return true }

    guard p.hasPrefix("*.") else {
        return p == h
    }

    // Wildcard: `*.example.com`
    let suffix = String(p.dropFirst(1)) // `.example.com`
    return h.count > suffix.count && h.hasSuffix(suffix)
}

/// Check if hostname is allowed by any of the domain patterns.
public func isAllowedDomain(patterns: [String], hostname: String) -> Bool {
    patterns.contains { domainMatches(pattern: $0, hostname: hostname) }
}

File: packages/shared-ts/src/secret-masking.ts
/**
 * Secret masking engine — HMAC-SHA256 based, format-preserving.
 *
 * Used on both agent side (env population, output scrubbing) and
 * server side (fetch proxy response scrubbing).
 */

// ---------- Known token prefixes ----------

const KNOWN_PREFIXES: string[] = [
  'ghp_', // GitHub PAT
  'gho_', // GitHub OAuth
  'ghu_', // GitHub user-to-server
  'ghs_', // GitHub server-to-server
  'ghr_', // GitHub refresh
  'github_pat_', // GitHub fine-grained PAT
  'sk-', // OpenAI / Stripe secret key
  'pk-', // Stripe publishable key
  'xoxb-', // Slack bot token
  'xoxp-', // Slack user token
  'xoxa-', // Slack app token
  'xoxs-', // Slack session token
  'AKIA', // AWS access key ID
  'ABIA', // AWS STS
  'ACCA', // AWS alternate
  'ASIA', // AWS temporary
  'sk-ant-', // Anthropic
  'Bearer ', // generic bearer (keep space)
];

// Sort longest-first so we match the most specific prefix
const SORTED_PREFIXES = [...KNOWN_PREFIXES].sort((a, b) => b.length - a.length);

/**
 * Detect a known prefix from the value. Returns the prefix string or empty.
 */
function detectPrefix(value: string): string {
  for (const p of SORTED_PREFIXES) {
    if (value.startsWith(p)) return p;
  }
  return '';
}

// ---------- HMAC-SHA256 ----------

async function hmacSha256(key: string, message: string): Promise<Uint8Array> {
  const enc = new TextEncoder();
  const cryptoKey = await crypto.subtle.importKey(
    'raw',
    enc.encode(key),
    { name: 'HMAC', hash: 'SHA-256' },
    false,
    ['sign']
  );
  const sig = await crypto.subtle.sign('HMAC', cryptoKey, enc.encode(message));
  return new Uint8Array(sig);
}

function toHex(bytes: Uint8Array): string {
  return Array.from(bytes)
    .map((b) => b.toString(16).padStart(2, '0'))
    .join('');
}

// ---------- Public API ----------

/**
 * Produce a deterministic, format-preserving masked value.
 *
 * `masked = prefix + hex(HMAC-SHA256(sessionId+secretName, realValue))`
 * truncated/repeated to match `realValue.length - prefix.length`.
 */
export async function mask(
  sessionId: string,
  secretName: string,
  realValue: string
): Promise<string> {
  const prefix = detectPrefix(realValue);
  const remainder = realValue.slice(prefix.length);

  const hmac = await hmacSha256(sessionId + secretName, realValue);
  let hex = toHex(hmac);

  // Repeat hex if remainder is longer than 64 hex chars
  while (hex.length < remainder.length) hex += hex;
  const maskedRemainder = hex.slice(0, remainder.length);

  return prefix + maskedRemainder;
}

export interface SecretPair {
  realValue: string;
  maskedValue: string;
}

/**
 * Build a reusable scrubber function that replaces every occurrence
 * of any `realValue` with its `maskedValue`.
 *
 * For a small number of secrets a simple sequential replace is fine.
 * Secrets are sorted longest-first to avoid partial-match issues.
 */
export function buildScrubber(secrets: SecretPair[]): (text: string) => string {
  if (secrets.length === 0) return (t) => t;

  // Sort longest real values first to avoid sub-string clobbering
  const sorted = [...secrets].sort((a, b) => b.realValue.length - a.realValue.length);

  return (text: string): string => {
    let result = text;
    for (const { realValue, maskedValue } of sorted) {
      if (result.includes(realValue)) {
        result = result.split(realValue).join(maskedValue);
      }
    }
    return result;
  };
}

/**
 * Domain glob matching.
 *
 * - `api.github.com` matches exactly `api.github.com`
 * - `*.github.com`   matches `api.github.com`, `uploads.github.com`,
 *                     but NOT `github.com` itself
 */
export function domainMatches(pattern: string, hostname: string): boolean {
  const p = pattern.toLowerCase();
  const h = hostname.toLowerCase();

  if (p === '*') return true;

  if (!p.startsWith('*.')) {
    return p === h;
  }

  // Wildcard: `*.example.com`
  const suffix = p.slice(1); // `.example.com`
  // hostname must end with `.example.com` AND have at least one char before
  return h.length > suffix.length && h.endsWith(suffix);
}

/**
 * Check if hostname is allowed by any of the domain patterns.
 */
export function isAllowedDomain(patterns: string[], hostname: string): boolean {
  return patterns.some((p) => domainMatches(p, hostname));
}

/**
 * Compatibility alias for node-server's historical name + arg order.
 * Prefer `isAllowedDomain(patterns, hostname)` in new code.
 */
export function matchesDomains(hostname: string, patterns: string[]): boolean {
  return isAllowedDomain(patterns, hostname);
}
